rst: add missing line/column info for some warnings

a-mr opened the pull request:
Some rst node types are post-processed after parsing, that is in either `resolveSubs` or in `rstgen.nim`, while lines/columns are stored only in tokens during parsing stage. There was no way to print proper  line/column for the corresponding warnings or errors. Now for all those 7 node kinds the line/column info is saved in a dedicated field `PRstNode.loc`.

Fixes bug 8 from #17340 .

a-mr committed:
rst: add missing line/column info for some warnings

a-mr committed:
add workaround

Araq commented on lib/packages/docutils/rstast.nim:
```diff
@@ -77,6 +77,11 @@ type
   PRstNode* = ref RstNode    ## an RST node
   RstNodeSeq* = seq[PRstNode]
+  PRstLocation* = ref object ## location inside a file
+    filename*: string
```
I would try to reuse the `TLineInfo` mechanism.

timotheecour replied:
see also `NimLineInfo` refs https://github.com/nim-lang/RFCs/issues/387

a-mr committed:
use TLineInfo/FileIndex for storing file names

a-mr committed:
fix blank lines in include file (rm harmful strip)

Araq commented on lib/packages/docutils/rst.nim:
```diff
@@ -589,32 +609,38 @@ proc newRstSharedState*(options: RstParseOptions,
+  s.msgHandler(s.currFilename, LineRstInit, ColRstInit, msgKind, arg)
+
+proc rstMessage*(files: seq[string], f: MsgHandler,
+                 li: ref TLineInfo | TLineInfo,
```
Why would it ever be a `ref TLineInfo`? Don't allocate these small objects on the heap.

a-mr replied:
I forgot that they are that small. OK,  done.

a-mr committed:
don't use ref TLineInfo

a-mr committed:
return `hasToc` as output parameter for uniformity

timotheecour commented on compiler/docgen.nim:
```diff
@@ -202,20 +201,26 @@ proc getOutFile2(conf: ConfigRef; filename: RelativeFile,
 proc newDocumentor*(filename: AbsoluteFile; cache: IdentCache; conf: ConfigRef,
-                    outExt: string = HtmlExt, module: PSym = nil): PDoc =
+                    outExt: string = HtmlExt, module: PSym = nil,
+                    isPureRst=false): PDoc =
```
```suggestion
                    isPureRst = false): PDoc =
```

timotheecour commented on compiler/docgen.nim:
```diff
@@ -202,20 +201,26 @@ proc getOutFile2(conf: ConfigRef; filename: RelativeFile,
-  const options = {roSupportRawDirective, roSupportMarkdown,
-                   roPreferMarkdown, roNimFile}
+  result.isPureRst = isPureRst
+  var options: RstParseOptions = {}
+  if isPureRst:
+    options = {roSupportRawDirective, roSupportMarkdown, roPreferMarkdown}
+  else:
+    options = {roSupportRawDirective, roSupportMarkdown, roPreferMarkdown,
+               roNimFile}
```
```suggestion
  var options= {roSupportRawDirective, roSupportMarkdown, roPreferMarkdown}
  if not isPureRst: options.incl roNimFile
```

timotheecour commented on compiler/docgen.nim:
```diff
@@ -1413,14 +1420,10 @@ proc commandDoc*(cache: IdentCache, conf: ConfigRef) =
-  var d = newDocumentor(filen, cache, conf, outExt)
-
-  d.isPureRst = true
+  var d = newDocumentor(filen, cache, conf, outExt, isPureRst=true)
```
```suggestion
  let d = newDocumentor(filen, cache, conf, outExt, isPureRst = true)
```

a-mr replied:
`let` does not work here because `d` is used as `var` in `finishGenerateDoc`.
I feel I will sometime go through the file and convert some `var` to `let` to satisfy your var-elimination urge :-)

timotheecour commented on compiler/lineinfos.nim:
```diff
@@ -195,6 +197,9 @@ const
   warnMax* = pred(hintSuccess)
   hintMin* = hintSuccess
   hintMax* = high(TMsgKind)
+  rstWarnings* = {warnRedefinitionOfLabel, warnUnknownSubstitutionX,
```
optional, and could be in future work: could all these be renamed with warnRst prefix for clarity?
this shouldn't impact users

a-mr replied:
may be, but separate warnings allow to turn each one to an error (or vice versa suppress it). And it's not that hard to maintain the list.

timotheecour replied:
no, i meant renaming them eg:
warnRedefinitionOfLabel => warnRstRedefinitionOfLabel
warnUnknownSubstitutionX => warnRstUnknownSubstitutionX
etc

a-mr replied:
makes sense, agree. Done for errors and warnings.

timotheecour commented on lib/packages/docutils/rst.nim:
```diff
@@ -509,7 +510,9 @@ type
     msgHandler: MsgHandler      # How to handle errors.
     findFile: FindFileHandler   # How to find files.
-    filename: string
+    files*: seq[string]         # map FileIndex -> file name (for storing
```
```suggestion
    filenames*: seq[string]         # map FileIndex -> file name (for storing
```
(optional nit)
file has C File connotation

timotheecour commented on lib/packages/docutils/rst.nim:
```diff
@@ -229,6 +229,7 @@ type
     meFootnoteMismatch = "mismatch in number of footnotes and their refs: $1",
     mwRedefinitionOfLabel = "redefinition of label '$1'",
```
add a comment: keep in snc with compiler/lineinfos.TMsgKind, and maybe re-order to make the common enums follow 1:1

a-mr replied:
done.

timotheecour commented on lib/packages/docutils/rst.nim:
```diff
@@ -579,6 +582,23 @@ proc whichRoleAux(sym: string): RstNodeKind =
+  if fid.int < files.len:
+    result = files[fid.int]
+  else:
+    result = "input"
```
what is `input` used for? is that same as `compiler/msgs.nim:177:3:  commandLineDesc* = "command line"` ?

a-mr replied:
just a fallback value, it should not happen. Alternatively a ValueError can be raised here.

timotheecour replied:
if it shouldn't happen than maybe use doAssert? (ie indicates program logic bug, not user error)

a-mr replied:
OK, switched to `doAssert`.

a-mr replied:
... and doAssert caught a bug :-)

timotheecour commented on tests/stdlib/trst.nim:
```diff
@@ -358,7 +406,7 @@ OtherStart
              :start-after: OtherStart
 """
-    doAssert "<em>Visible</em>" == rstTohtml(input, {}, defaultConfig())
+    check "<em>Visible</em> " == rstTohtml(input, {}, defaultConfig())
```
how about using strbasics.strip (inline strip) at the end of `rstTohtml` instead?

a-mr replied:
good idea. Applied it.

timotheecour commented on tests/stdlib/trst.nim:
```diff
@@ -340,6 +341,53 @@ suite "RST indentation":
+          rnRef
+            rnLeaf  'brokenLink'
+      """)
+    removeFile("other.rst")
```
* not DRY
* bad style to write in the source tree and with a generic name (eg what if you ^C testament and are left with un-clean git status, or worse, it overwrites another test file)
* instead use either:
```nim
import stdtest/specialpaths
const tmpFile = buildDir / "D20210706T170748.rst"
```
or `createTempFile` from std/tempfiles

a-mr replied:
I copy pasted it from this very file. Let us postpone it for `trst.nim` refactoring PR.

timotheecour commented on lib/packages/docutils/rstast.nim:
```diff
@@ -92,9 +96,15 @@ type
                               ## auto-numbered ones without a label)
+    of rnRef, rnSubstitutionReferences,
+        rnInterpretedText, rnField, rnInlineCode, rnCodeBlock:
+      li*: TLineInfo          ## To have line/column info for warnings at
```
`li` is neither standard in nim sources nor self documenting; instead use `info` which is at least standard in nim compiler

ditto elsewhere (eg PFootnoteRefInfo.li and in proc params eg `li: TLineInfo`)

a-mr replied:
thank you, done.

timotheecour commented on lib/packages/docutils/rstast.nim:
```diff
@@ -77,6 +77,10 @@ type
 
   PRstNode* = ref RstNode    ## an RST node
   RstNodeSeq* = seq[PRstNode]
+  PFootnoteRefInfo* = ref object ## location inside a file, only for
```
ref object seems like a false optimization; the cost of allocs will dwarf any savings here (not that those savings should matter much, but not using ref seems simpler here anyways)

I'm betting the total number of `RstNode` nodes generated is itself dwarfed by total number of PNode generated during nim doc anyways (and also shouldn't be a concern for rst2html)

a-mr replied:
may be. But araq previously complained about similar case.

timotheecour replied:
where? in fact https://github.com/nim-lang/Nim/pull/18383#discussion_r663351147 tells the same thing: 
> Why would it ever be a ref TLineInfo? Don't allocate these small objects on the heap.

but this can be addressed in followup PR, not a blocker, this PR is already big enough

a-mr replied:
In another PR. It was in a similar situation (when 1 extra field was proposed for addition) a few months ago.

a-mr replied:
actually saving `order` field is not necessary for `rnFootnoteRef`, instead I added 2 counters to `RstSharedState` that are processed in `resolveSubs`.

So the problem is gone without need to sacrifice any memory consumption.

@timotheecour , see the latest commit.

timotheecour replied:
ok!

a-mr committed:
Update compiler/docgen.nim

Co-authored-by: Timothee Cour <timothee.cour2@gmail.com>

a-mr committed:
Update compiler/docgen.nim

Co-authored-by: Timothee Cour <timothee.cour2@gmail.com>

a-mr committed:
Update lib/packages/docutils/rst.nim

Co-authored-by: Timothee Cour <timothee.cour2@gmail.com>

a-mr committed:
address review - stylistic things

a-mr committed:
Update compiler/docgen.nim

Co-authored-by: Timothee Cour <timothee.cour2@gmail.com>

a-mr committed:
unify RST warnings/errors names

timotheecour commented on compiler/lineinfos.nim:
```diff
@@ -32,13 +32,13 @@ type
     errXExpected,
-    errGridTableNotImplemented,
+    errRstGridTableNotImplemented,
     errMarkdownIllformedTable,
```
move up errMarkdownIllformedTable 1 spot? ABI compat isn't mandated

a-mr replied:
why? this order is the same in all the 3 files.

timotheecour replied:
to avoid splitting the `errRstxxx` entries with `errMarkdownIllformedTable` in the middle (more obvious when looking at the after-the-diff output)

a-mr replied:
changed to `errRstMarkdownIllFormadTable` for uniformity.

a-mr committed:
doAssert + minor name change

a-mr committed:
fix a bug caught by doAssert

a-mr committed:
apply strbasics.strip to final HTML/Latex

a-mr committed:
rm redundant filename

a-mr committed:
fix test after rebase

a-mr committed:
delete `order` from rnFootnoteRef,

also display errors/warnings properly when footnote references are from
different files

timotheecour commented on compiler/lineinfos.nim:
```diff
@@ -196,6 +200,9 @@ const
   warnMax* = pred(hintSuccess)
   hintMin* = hintSuccess
   hintMax* = high(TMsgKind)
+  rstWarnings* = {warnRstRedefinitionOfLabel, warnRstUnknownSubstitutionX,
+                  warnRstBrokenLink, warnRstLanguageXNotSupported,
+                  warnRstFieldXNotSupported, warnRstStyle}
```
```suggestion
  rstWarnings* = {warnRstRedefinitionOfLabel..warnRstStyle}
```
(or something similar)

timotheecour commented on tests/stdlib/trstgen.nim:
```diff
@@ -454,22 +454,23 @@ Some chapter
-    doAssert "<h2 id=\"another2\">Another2</h2>" in output10
-    doAssert "<h3 id=\"more3\"><center>More3</center></h3>" in output10
-
+    doAssert "<h1 id=\"level1\"><center>Level1</center></h1>" in output
```
how about using triple quoted strings instead so that it's easier to copy paste without quoting, eg:
```nim
doAssert "<h1 id=\"level1\"><center>Level1</center></h1>" in output
=>
doAssert """<h1 id="level1"><center>Level1</center></h1>""" in output
```
can be done in separate PR though

it could even be written as:
```nim
for line in """
<h1 id="level1"><center>Level1</center></h1>
<h2 id="level2">Level2</h2>
...
""":
  doAssert line in output
```

if you want to maintain ordering there's also `testutils.greedyOrderedSubsetLines`

timotheecour commented on tests/stdlib/trstgen.nim:
```diff
@@ -509,14 +511,14 @@ Some chapter
+    rstGenera.renderRstToOut(rst, output)
+    doAssert rstGenera.meta[metaTitle] == "Title0"
+    doAssert rstGenera.meta[metaSubTitle] == ""
+    doAssert output ==
```
use multiline literal instead (can be done in separate PR)

timotheecour commented on tests/stdlib/trstgen.nim:
```diff
@@ -628,7 +630,7 @@ let x = 1
     let p3 = """<p>Par3 <tt class="docutils literal"><span class="pre">""" & id"value3" & "</span></tt>.</p>"
     let p4 = """<p>Par4 <tt class="docutils literal"><span class="pre">value4</span></tt>.</p>"""
-    let expected = p1 & p2 & "\n" & p3 & "\n" & p4 & "\n"
+    let expected = p1 & p2 & "\n" & p3 & "\n" & p4
```
ditto, multiline literal is much clearer than concatenating like this

timotheecour commented on tests/stdlib/trstgen.nim:
```diff
@@ -1144,10 +1145,33 @@ Test1
                      "extra arguments were given to number-lines: ' let a = 1'")
     check "" == output
 
+  test "code-block warning":
```
not a blocker because the situation after this PR improves in all cases, but the line numbers and columns are off by ~1 for nested doc comments, eg:

before PR: all lines/cols are 1 (useless)
after PR: main3 is correct, main1 and main2 are off by ~1

`````nim

proc main1*()=
  ##[
    .. code:: Nim
       :unsupportedField: anything
    .. code:: unsupportedLang
       anything
    ```anotherLang
    someCode
  ```
  ]##
  discard


proc main2*()=
  ##[
  .. code:: Nim
     :unsupportedField: anything
  .. code:: unsupportedLang
     anything
  ```anotherLang
  someCode
  ```
  ]##
  discard

proc main3*()=
  ## .. code:: Nim
  ##    :unsupportedField: anything
  ## .. code:: unsupportedLang
  ##    anything
  ## ```anotherLang
  ## someCode
  ## ```
  discard
`````

a-mr replied:
yes, line info is saved only for first `#` of `##`, not for real comment inside.
We talked about this problem in https://github.com/timotheecour/Nim/issues/658

timotheecour commented on lib/packages/docutils/rstast.nim:
```diff
@@ -9,7 +9,7 @@
 ## This module implements an AST for the `reStructuredText`:idx: parser.
 
-import strutils, json
+import strutils, json, ../../../compiler/lineinfos
```
```suggestion
import strutils, json, "$lib/../compiler/lineinfos"
```

note: this would be wrong as it wouldn't work when bootstrapping from an external nim with --lib:lib
```
import strutils, json, "$nim/compiler/lineinfos"
```

timotheecour commented on lib/packages/docutils/rstast.nim:
```diff
@@ -107,6 +111,12 @@ proc len*(n: PRstNode): int =
+
+proc newRstNode*(kind: RstNodeKind, info: TLineInfo,
+                 sons: seq[PRstNode] = @[]): PRstNode =
+  result = PRstNode(kind: kind, sons: sons)
+  result.info = info
```
```suggestion
  result = PRstNode(kind: kind, sons: sons, info: info)
```

timotheecour commented on lib/packages/docutils/rstast.nim:
```diff
@@ -107,6 +111,12 @@ proc len*(n: PRstNode): int =
 proc newRstNode*(kind: RstNodeKind, sons: seq[PRstNode] = @[],
                  anchor = ""): PRstNode =
   result = PRstNode(kind: kind, sons: sons)
+  result.anchor = anchor
```
```suggestion
  result = PRstNode(kind: kind, sons: sons, anchor: anchor)
```

timotheecour commented on lib/packages/docutils/rstgen.nim:
```diff
@@ -40,7 +40,7 @@
 
 import strutils, os, hashes, strtabs, rstast, rst, highlite, tables, sequtils,
-  algorithm, parseutils
+  algorithm, parseutils, ../../../compiler/lineinfos, std/strbasics
```
```suggestion
  algorithm, parseutils, "$lib/../compiler/lineinfos", std/strbasics
```

timotheecour commented on lib/packages/docutils/rst.nim:
```diff
@@ -196,7 +196,7 @@
 import
   os, strutils, rstast, std/enumutils, algorithm, lists, sequtils,
-  std/private/miscdollars
+  std/private/miscdollars, ../../../compiler/lineinfos
```
```suggestion
  std/private/miscdollars, "$lib../compiler/lineinfos"
```

a-mr replied:
ok, done, thank you.

timotheecour commented on lib/packages/docutils/rst.nim:
```diff
@@ -501,15 +503,19 @@ type
     msgHandler: MsgHandler      # How to handle errors.
     findFile: FindFileHandler   # How to find files.
-    filename: string
+    filenames*: seq[string]     # map FileIndex -> file name (for storing
```
if it's a map, shouldn't this be a Table? this would also simplify setCurrFilename , getFilename etc

a-mr replied:
I'm fine with both ways, readability is the same.
- `seq[string]` should be faster for typical cases when there is only 1 file
- `Table` will be faster when there is >~ 10 files included

timotheecour replied:
I don't think performance is any concern here (not a bottleneck), but it seemed to me Table would simplify code, but again, can be deferred to future work

timotheecour commented on lib/packages/docutils/rst.nim:
```diff
@@ -589,32 +611,37 @@ proc newRstSharedState*(options: RstParseOptions,
+
+proc rstMessage*(filenames: seq[string], f: MsgHandler,
+                 info: TLineInfo, msgKind: MsgKind, arg: string) =
+  ## Print warnings using `info: TLineInfo`, i.e. in 2nd-pass warnings for
```
```suggestion
  ## Print warnings using `info`, i.e. in 2nd-pass warnings for
```

timotheecour reviewed (approved):
LGTM, remaining comments are not critical and can be addressed in a followup PR so that this PR doesn't bitrot

a-mr committed:
Update compiler/lineinfos.nim

Co-authored-by: Timothee Cour <timothee.cour2@gmail.com>

a-mr committed:
Update lib/packages/docutils/rstast.nim

Co-authored-by: Timothee Cour <timothee.cour2@gmail.com>

a-mr committed:
Update lib/packages/docutils/rstast.nim

Co-authored-by: Timothee Cour <timothee.cour2@gmail.com>

a-mr committed:
Update lib/packages/docutils/rstast.nim

Co-authored-by: Timothee Cour <timothee.cour2@gmail.com>

a-mr committed:
revert because of error:

Error: cannot prove that it's safe to initialize 'info' with the runtime value for the discriminator 'kind'

a-mr committed:
Update lib/packages/docutils/rstgen.nim

Co-authored-by: Timothee Cour <timothee.cour2@gmail.com>

a-mr committed:
apply suggestion

a-mr committed:
Update lib/packages/docutils/rst.nim

Co-authored-by: Timothee Cour <timothee.cour2@gmail.com>

timotheecour reviewed (approved):
still LGTM (remaing comments can be addressed in followup PR)

a-mr committed:
add Table for string->file name mapping

narimiran committed:
do not import compiler/lineinfos

narimiran committed:
fix ambiguous calls

narimiran commented:
@a-mr I've pushed two commits, which remove imports of `compiler/lineinfos` (per @Araq's request). I wanted to let you know, so if you decide to push something more, to pull first.
